File: src/python/DataEvent.py
import threading
import queue
import time
import asyncio
from collections import deque
import traceback
import sys

class DataEvent(threading.Event):
    def __init__(self, shouldHop=False):
        super().__init__();
        self.shouldHop = shouldHop
        self.q = deque()
        self.lock = threading.Lock()

    def set(self, data=None):
        self.lock.acquire()
        if self.shouldHop and len(self.q) > 0:
            self.q[0] = data
        else:     
            self.q.append(data)
        self.lock.release()
        super().set()

    def waitRun(self, f):
        super().wait()
        self.lock.acquire()
        q_ = self.q.copy()
        self.q.clear()
        self.lock.release()
        for d in q_:
            try:
                f(d)
            except Exception as ex:
                print(ex)
                traceback.print_exc(file=sys.stdout)


class AsyncDataEvent(asyncio.Event):
    def __init__(self, shouldHop=False):
        super().__init__();
        self.shouldHop = shouldHop
        self.q = deque()
        self.lock = threading.Lock()

    def set(self, data=None):
        self.lock.acquire()
        try:
            if self.shouldHop and len(self.q) > 0:
                self.q[0] = data
            else:
                self.q.append(data)
        except Exception as ex:
            print(ex)
            traceback.print_exc(file=sys.stdout)
        self.lock.release()
        super().set()

    async def waitRun(self, f):
        await super().wait()
        self.lock.acquire()
        try:
            q_ = self.q.copy()
            self.q.clear()
        except Exception as ex:
            print(ex)
            traceback.print_exc(file=sys.stdout)
        self.lock.release()
        await self.run_sync(q_, f)

    async def run_sync(self, data, f):
        for d in data:
            try:
                await f(d)
            except Exception as ex:
                print(ex)
                printf(f'   in: {f}')
                traceback.print_exc(file=sys.stdout)

# class AsyncDataEventVerbose(asyncio.Event):
#     def __init__(self, shouldHop=False, txt=None):
#         super().__init__();
#         self.shouldHop = shouldHop
#         self.q = deque()
#         self.lock = threading.Lock()
#         self.txt = txt

#     def set(self, data=None):
#         print(f"   acquiring set lock {self}")
#         self.lock.acquire()
#         print(f"   acquired set lock {self}")
#         if self.shouldHop and len(self.q) > 0:
#             self.q[0] = data
#         else:     
#             self.q.append(data)
#         self.lock.release()
#         print(f"   released set lock {self}")
#         # self._loop.call_soon_threadsafe(super().set)
#         super().set()
#         print(f"   super set {self}  {asyncio.get_running_loop()._thread_id}")

#     async def waitRun(self, f):
#         print(f"   waiting {self} {asyncio.get_running_loop()._thread_id}")
#         await super().wait()
#         print(f"   super out of wait {self}")
#         self.lock.acquire()
#         q_ = self.q.copy()
#         self.q.clear()
#         self.lock.release()
#         print(f'Data: {q_} f:{f}')
#         await self.run_sync(q_, f)

#     async def run_sync(self, data, f):
#         for d in data:
#             print(f'Running {f} on {d}')
#             try:
#                 await f(d)
#             except Exception as ex:
#                 print(ex)
#                 traceback.print_exc(file=sys.stdout)
#             print(f'Ran {f} on {d}')

File: src/components/DataEvent.js
import { EventEmitter } from 'events';

export class DataEvent extends EventEmitter {
    constructor(shouldHop = false) {
        super();
        this.shouldHop = shouldHop;
        this.q = [];
        this.lock = false;
    }

    set(data = null) {
        this.lock = true;
        if (this.shouldHop && this.q.length > 0) {
            this.q[0] = data;
        } else {
            this.q.push(data);
        }
        this.lock = false;
        this.emit('event');
    }

    waitRun(f) {
        this.once('event', () => {
            const q_ = [...this.q];
            this.q = [];
            for (const d of q_) {
                try {
                    f(d);
                } catch (ex) {
                    console.error(ex);
                    console.trace();
                }
            }
        });
    }
}

export class AsyncDataEvent extends EventEmitter {
    constructor(shouldHop = false, verbose = false) {
        super();
        this.shouldHop = shouldHop;
        this.verbose = verbose;
        this.q = [];
        this.lock = false;
    }

    set(data = null) {
        this.lock = true;
        try {
            if (this.shouldHop && this.q.length > 0) {
                this.q[0] = data;
            } else {
                this.q.push(data);
            }
            if (this.verbose) console.log("AsyncDataEvent set: ", this.q);
        } catch (ex) {
            console.error(ex);
            console.trace();
        }
        this.lock = false;
        this.emit('event');
    }

    async waitRun(f) {
        if (this.verbose) console.log("AsyncDataEvent waiting");
        await new Promise(resolve => this.once('event', resolve));
        if (this.verbose) console.log("AsyncDataEvent out of wait: ", this.q);
        this.lock = true;
        try {
            const q_ = [...this.q];
            this.q = [];
            this.run(q_, f);
        } catch (ex) {
            console.error(ex);
            console.trace();
        }
        this.lock = false;
    }

    async run(data, f) {
        if (this.verbose) console.log("AsyncDataEvent running: ", f, data);
        for (const d of data) {
            try {
                await f(d);
            } catch (ex) {
                console.error(ex);
                console.log(`   in: ${f}`);
                console.trace();
            }
        }
    }
}

